crates/holochain/src/core/workflow/incoming_dht_ops_workflow/incoming_ops_batch.rs

```diff
@@ -0,0 +1,146 @@
+use crate::core::workflow::error::WorkflowResult;
+use holo_hash::DhtOpHash;
+use holochain_types::dht_op::DhtOp;
+use holochain_types::prelude::DhtOpHashed;
+use holochain_types::share::RwShare;
+
+type InOpBatchSnd = tokio::sync::oneshot::Sender<WorkflowResult<()>>;
+type InOpBatchRcv = tokio::sync::oneshot::Receiver<WorkflowResult<()>>;
+
+#[derive(Debug)]
+pub struct InOpBatchEntry {
+    pub snd: InOpBatchSnd,
+    pub request_validation_receipt: bool,
+    pub ops: Vec<DhtOpHashed>,
+}
+
+/// A batch of incoming ops memory.
+#[derive(Clone)]
+pub struct IncomingOpsBatch(RwShare<InOpBatch>);
+
+impl IncomingOpsBatch {
+    pub fn is_running(&self) -> bool {
+        self.0.share_ref(|b| b.is_running)
+    }
+
+    /// if result.0.is_none() -- we queued it to send later
+    /// if result.0.is_some() -- the batch should be run now
+    pub fn check_insert(
+        &self,
+        request_validation_receipt: bool,
+        ops: Vec<DhtOpHashed>,
+    ) -> (Option<Vec<InOpBatchEntry>>, InOpBatchRcv) {
+        let (snd, rcv) = tokio::sync::oneshot::channel();
+        let entry = InOpBatchEntry {
+            snd,
+            request_validation_receipt,
+            ops,
+        };
+        self.0.share_mut(|batch| {
+            if batch.is_running {
+                // there is already a batch running, just queue this
+                batch.pending.push(entry);
+                (None, rcv)
+            } else {
+                // no batch running, run this (and assert we never collect stragglers)
+                assert!(batch.pending.is_empty());
```

Comment:
All this code is just moved. My only change is to make these `impl` functions since they took an `&self` as their first argument anyway.

Do we want to keep these assertions? This one should not be reachable based on my analysis of the code but the other one IS reachable as a test demonstrates. It seems unnecessary to leave these in now the code is tested.

Reply:
Maybe just change to `debug_assert!`? Can't hurt, nice to have.